src/Uno.UI/UI/Xaml/Media/VisualTreeHelper.cs

```diff
@@ -193,5 +202,326 @@ public static FrameworkElement TryAdaptNative(_View view)
+
+				// If we found a stale element in child sub-tree, keep it and stop looking for stale elements
+				if (childResult.stale is { })
+				{
+					stale = childResult.stale;
+					isChildStale = null;
+				}
+
+				// If we found an acceptable element in the child's sub-tree, job is done!
+				if (childResult.element is { })
+				{
+					if (isChildStale is { }) // Also indicates that stale is null
+					{
+						// If we didn't find any stale root in previous children or in the child's sub tree,
+						// we continue to enumerate sibling children to detect a potential stale root.
+
+						while (child.MoveNext())
+						{
+							if (isChildStale(child.Current))
+							{
+								stale = SearchDownForStaleBranch(child.Current!, isChildStale);
+								break;
+							}
+						}
+					}
+
+					TRACE($"> found child: {childResult.element.GetDebugName()} | stale branch: {stale?.ToString() ?? "-- none --"}");
+					return (childResult.element, stale);
+				}
+			}
+
+			// We didn't find any child at the given position, validate that element can be touched (i.e. not HitTestability.Invisible),
+			// and the position is in actual bounds (which might be different than the clipping bounds)
+			if (elementHitTestVisibility == HitTestability.Visible && renderingBounds.Contains(posRelToElement))
+			{
+				TRACE($"> LEAF! ({element.GetDebugName()} is the OriginalSource) | stale branch: {stale?.ToString() ?? "-- none --"}");
+				return (element, stale);
+			}
+			else
+			{
+				// If no stale element found yet, validate if the current is stale.
+				// Note: no needs to search down for stale child, we already did it!
+				if (isStale?.Invoke(element) ?? false)
+				{
+					stale = new Branch(element, stale?.Leaf ?? element);
+				}
+
+				TRACE($"> NOT FOUND (HitTestability.Invisible or out of the **render** bounds) | stale branch: {stale?.ToString() ?? "-- none --"}");
+				return (default, stale);
+			}
+		}
+
+		private static Branch SearchDownForStaleBranch(UIElement staleRoot, Predicate<UIElement> isStale)
+			=> new Branch(staleRoot, SearchDownForStaleLeaf(staleRoot, isStale));
+
+		private static UIElement SearchDownForStaleLeaf(UIElement staleRoot, Predicate<UIElement> isStale)
+		{
+			foreach (var child in staleRoot.GetChildren().OfType<UIElement>().Reverse())
+			{
+				if (isStale(child))
+				{
+					return SearchDownForStaleLeaf(child, isStale);
+				}
+			}
+
+			return staleRoot;
+		}
+
+		#region Helpers
+		private static Func<IEnumerable<UIElement>, IEnumerable<UIElement>> Except(UIElement element)
+			=> children => children.Except(element);
+
+		private static Func<IEnumerable<UIElement>, IEnumerable<UIElement>> SkipUntil(UIElement element)
+			=> children => SkipUntilCore(element, children);
+
+		private static IEnumerable<UIElement> SkipUntilCore(UIElement element, IEnumerable<UIElement> children)
+		{
+			using var enumerator = children.GetEnumerator();
+			while (enumerator.MoveNext() && enumerator.Current != element)
+			{
```

Comment:
Codacy found an issue: [Either remove or fill this block of code.](https://app.codacy.com/gh/unoplatform/uno/pullRequest?prid=6334034)